From 71a04fb7f990677db3673bb768638484264fe807 Mon Sep 17 00:00:00 2001
From: Andrew Wang <andrew.wang@cloudera.com>
Date: Tue, 26 May 2015 11:55:02 -0700
Subject: [PATCH 576/596] CLOUDERA-BUILD. CDH-28019. Cross site scripting
 vulnerability in dfsnodelist.jsp.

(cherry picked from commit 2f5e6d05627eea556ec319e84db62dd742febe9d)
---
 .../hdfs/server/namenode/NamenodeJspHelper.java    |   18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeJspHelper.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeJspHelper.java
index 5a25d4f..2eeb447 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeJspHelper.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeJspHelper.java
@@ -40,6 +40,7 @@
 import javax.servlet.http.HttpServletResponse;
 import javax.servlet.jsp.JspWriter;
 
+import org.apache.commons.lang.StringEscapeUtils;
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.ha.HAServiceProtocol.HAServiceState;
 import org.apache.hadoop.hdfs.DFSConfigKeys;
@@ -878,8 +879,21 @@ void generateNodesList(ServletContext context, JspWriter out,
       sorterOrder = request.getParameter("sorter/order");
       if (sorterField == null)
         sorterField = "name";
-      if (sorterOrder == null)
-        sorterOrder = "ASC";
+
+      String[] validSortOrders = new String[]{"ASC", "DSC"};
+      String validatedOrder = null;
+      for (String order : validSortOrders) {
+        if (sorterOrder.equals(order)) {
+          validatedOrder = order;
+          break;
+        }
+      }
+      if (validatedOrder == null) {
+        out.print("Invalid order: ");
+        out.print(StringEscapeUtils.escapeHtml(sorterOrder));
+        return;
+      }
+      sorterOrder = validatedOrder;
 
       JspHelper.sortNodeList(live, sorterField, sorterOrder);
 
-- 
1.7.9.5

